packages/fixie/src/agent.ts

```diff
@@ -317,30 +345,213 @@ export class FixieAgent {
+    const subProcess = spawn(argv0, args, { cwd: agentPath, signal });
+    subProcess.stdout.on('data', (sdata: string) => {
+      console.log(`🌱 Agent stdout: ${sdata}`);
+    });
+    subProcess.stderr.on('data', (sdata: string) => {
+      console.error(`🌱 Agent stdout: ${sdata}`);
+    });
+    subProcess.on('close', (returnCode: number) => {
+      term(`🌱 Agent child process exited with code ${returnCode} - `).red('exiting.\n');
+    });
+    return subProcess;
+  }
+
+  /** Deploy an agent from the given directory. */
+  public static async DeployAgent(
+    client: FixieClient,
+    agentPath: string,
+    environmentVariables: Record<string, string> = {}
+  ): Promise<string> {
+    const config = await FixieAgent.LoadConfig(agentPath);
+    const agentId = `${(await client.userInfo()).username}/${config.handle}`;
+    term('🦊 Deploying agent ').green(agentId)('...\n');
+
+    // Check that the package.json path exists in this directory.
+    if (!fs.existsSync(path.resolve(path.join(agentPath, 'package.json')))) {
+      throw Error(`No package.json found in ${agentPath}. Only JS-based agents are supported.`);
+    }
+    const agent = await this.ensureAgent(client, agentId, config);
     const tarball = FixieAgent.getCodePackage(agentPath);
     const spinner = ora(' 🚀 Deploying...').start();
-    const revision = await agent.createRevision(tarball, environmentVariables);
+    const revision = await agent.createRevision({ tarball, environmentVariables });
     spinner.succeed(`Revision ${revision} was deployed to ${agent.agentUrl()}`);
     return revision;
   }
 
-  /** Get the current agent revision. */
-  public async getCurrentRevision(): Promise<AgentRevision> {
-    const result = await this.client.gqlClient().query({
-      fetchPolicy: 'no-cache',
-      query: gql`
-        query GetRevisionId($agentId: String!) {
-          agentById(agentId: $agentId) {
-            currentRevision {
-              id
-              created
+  /** Run an agent locally from the given directory. */
+  public static async ServeAgent({
+    client,
+    agentPath,
+    tunnel,
+    port,
+    environmentVariables,
+    debug,
+  }: {
+    client: FixieClient;
+    agentPath: string;
+    tunnel?: boolean;
+    port: number;
+    environmentVariables: Record<string, string>;
+    debug?: boolean;
+  }) {
+    const config = await FixieAgent.LoadConfig(agentPath);
+    const agentId = `${(await client.userInfo()).username}/${config.handle}`;
+    term('🦊 Serving agent ').green(agentId)('...\n');
+
+    // Check if the package.json path exists in this directory.
+    if (!fs.existsSync(`${agentPath}/package.json`)) {
+      throw Error(`No package.json found in ${agentPath}. Only JS-based agents are supported.`);
+    }
+
+    // Start the agent process locally.
+    const agentProcess = FixieAgent.spawnAgentProcess(agentPath, port);
+
+    // Wait for 5 seconds for it to start up.
+    await new Promise((resolve) => setTimeout(resolve, 5000));
```

Comment:
This can be a follow-up, but it's much better to do a spinning assert – sleeps are very brittle.

Reply:
As you can tell, I Have No Idea What I'm Doing[tm], so let me know what you mean by this :-)

Reply:
I mean doing something like:

```ts
async function waitForServerStart() {
    return new Promise(resolve => {
        while (timeoutNotYetExceeded) {
            if (serverIsUp()) { resolve(); return; }
        }
        throw new Error(`Server didn't start within ${timeoutMs} milliseconds. You could try /* ... */`)
    });
}
```

Reply:
I agree we should do something smarter here, but do you mind if we do it as a follow up? I would need to write more code to figure out if the server is up, etc.